### DISCUSSION TITLE
Slowness problem when creating records

### DISCUSSION THREAD

**Discussion by Abel1011** [planetscale/discussion#269] [Bug Reports] [answered]
Good day to everyone, and thank you for your fantastic service. I am working on a project and in general, it goes very fast for individual queries and creations, but it takes a very long time (not always) when I create group records.

I am using prisma and Next.js deployed in VERCEL and I use Axiom for the records, and several times I got this error:

PrismaClientUnknownRequestError: 
Invalid `prisma.tokenUsage.createMany()` invocation:

Error in connector: Error querying the database: Server error: `ERROR 70100 (1317): vttablet: rpc error: code = Aborted desc = transaction 1662016673348263063: ended at 2022-09-03 15:49:00.322 UTC (exceeded timeout: 20s) (CallerID: planetscale-admin)'
    at RequestHandler.handleRequestError (/var/task/node_modules/@prisma/client/runtime/index.js:29758:13)
    at RequestHandler.request (/var/task/node_modules/@prisma/client/runtime/index.js:29737:12)
    at async /var/task/.next/server/chunks/3696.js:1046:20
    at async PrismaClient._request (/var/task/node_modules/@prisma/client/runtime/index.js:30709:16)
    at async Promise.all (index 1)
    at async longFormCompletion (/var/task/.next/server/pages/api/completions/long-form.js:242:9)
    at async Object.apiResolver (/var/task/node_modules/next/dist/server/api-utils/node.js:184:9)
    at async NextNodeServer.runApi (/var/task/node_modules/next/dist/server/next-server.js:403:9)
    at async Object.fn (/var/task/node_modules/next/dist/server/base-server.js:493:37)
    at async Router.execute (/var/task/node_modules/next/dist/server/router.js:222:36) {
  clientVersion: '4.3.0'
}

Through the prisma logs I was able to verify that the queries take quite a long time, even if the grouped records are only 3 or 4. Some results were:

Query TokenUsage.createMany took 63444ms (1 record)
Query TokenUsage.createMany took 181053ms (2 record)
Query Record.createMany took 180955ms (3 record)

These are some of the results I got, but the strange thing is that it doesn't always take this long. Occasionally, I get these values:

Query TokenUsage.createMany took 95ms (3 record).
Query Record.createMany took 79ms (2 records)
Query Record.createMany took 56ms (3 records)

No other query has this problem (only createMany). In that sense, I wanted to know if there is a problem with that type of query in Planetscale or is it related to the capability they provide at the Hobby level.

In any case, do you recommend me to create the data individually using a for instead of a createMany?, as there are usually few records to be made (at most 16).

Thank you in advance for any guidance you can give me. Best regards

**Comment by Abel1011** [comment]
Any advice, recommendations or help to debug the source of this problem is welcome.

**Comment by jgreet** [comment, marked as answer]
It might be useful to look at what happens during those `createMany` calls. Are they ever waiting on calls to other external systems, or performing some other operation that could take a varying amount of time?

Inserting multiple records shouldn't be a problem but if the entire createMany operation happens within a transaction then it can fail with the error you mentioned if the transaction isn't committed or rolled back within 20 seconds, even if the actual `INSERT` statements finish quickly.

**Reply by Abel1011** [reply]
Hello, the workflow in the API path is as follows:

-A database call is made to get user information.

-With the user information, the call to an external api is made.

-After receiving the data returned by the external api it is saved to the database with createMany. 

To perform the time logging I am using the following code:

```
prisma.$use(async (params, next) => {
  
  const before = Date.now();

  const result = await next(params);

  const after = Date.now();

  if (params.action === 'createMany') {
    console.log(`Query ${params.model}.${params.action} took ${after - before}ms`,params.args)
  }

  return result
})
```

I understand that this function gives me the time it takes just to register the createMany transaction; or maybe it is considering the time from the first call to the database?

```
08 sept, 10:49:26
Query TokenUsage.createMany took 154ms {
  data: [
    {
      count: 38,
      type: 'prompt',
      createdAt: 2022-09-08T15:49:19.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 30,
      cost: 0.00076,
      aiGenerationId: 'cmpl-5oE4FBiO1SUV83qcRSdiYYYtl0f8c',
      user_id: 147,
      template_id: 'TfIiAxvbyK4NaagdxcQzb'
    },
    {
      count: 195,
      type: 'completion',
      createdAt: 2022-09-08T15:49:19.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 156,
      cost: 0.0039,
      aiGenerationId: 'cmpl-5oE4FBiO1SUV83qcRSdiYYYtl0f8c',
      user_id: 147,
      template_id: 'TfIiAxvbyK4NaagdxcQzb'
    }
  ]
}

08 sept, 10:46:54
Query TokenUsage.createMany took 62521ms {
  data: [
    {
      count: 38,
      type: 'prompt',
      createdAt: 2022-09-08T15:45:43.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 30,
      cost: 0.00076,
      aiGenerationId: 'cmpl-5oE0lw3cTOXmJKIHJvgMtSdPIDIFy',
      user_id: 147,
      template_id: 'TfIiAxvbyK4NaagdxcQzb'
    },
    {
      count: 157,
      type: 'completion',
      createdAt: 2022-09-08T15:45:43.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 113,
      cost: 0.00314,
      aiGenerationId: 'cmpl-5oE0lw3cTOXmJKIHJvgMtSdPIDIFy',
      user_id: 147,
      template_id: 'TfIiAxvbyK4NaagdxcQzb'
    }
  ]
}


08 sept, 10:22:07
Query TokenUsage.createMany took 230440ms {
  data: [
    {
      count: 193,
      type: 'prompt',
      createdAt: 2022-09-08T15:18:11.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 152,
      cost: 0.00386,
      aiGenerationId: 'cmpl-5oDa7njAO1HFHe853R2dqMJmoCdiN',
      user_id: 1,
      template_id: 'long-form'
    },
    {
      count: 150,
      type: 'completion',
      createdAt: 2022-09-08T15:18:11.000Z,
      model: 'text-davinci-002',
      aiEnterprise: 'openai',
      words: 115,
      cost: 0.003,
      aiGenerationId: 'cmpl-5oDa7njAO1HFHe853R2dqMJmoCdiN',
      user_id: 1,
      template_id: 'long-form'
    }
  ]
}

08 sept, 08:48:46
Query TokenUsage.createMany took 3988ms {
  data: [
    {
      count: 38,
      type: 'prompt',
      createdAt: 2022-09-08T13:48:42.741Z,
      model: 'j1-grande',
      aiEnterprise: 'ai21labs',
      words: 46,
      cost: 0.0008,
      aiGenerationId: 'b1488f95-5c30-4fca-880f-1cedbc124636',
      user_id: 110,
      template_id: 'long-form'
    },
    {
      count: 1,
      type: 'completion',
      createdAt: 2022-09-08T13:48:42.741Z,
      model: 'j1-grande',
      aiEnterprise: 'ai21labs',
      words: 0,
      cost: 0.00008,
      aiGenerationId: 'b1488f95-5c30-4fca-880f-1cedbc124636',
      user_id: 110,
      template_id: 'long-form'
    }
  ]
}

```

**Reply by jgreet** [reply]
>I understand that this function gives me the time it takes just to register the createMany transaction; or maybe it is considering the time from the first call to the database?

I think you're doing it right but I'm not a Prisma expert 😄 . One other thing that might help is enabling [query logging](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#log-to-stdout) which could help you see anything else going on between the `BEGIN` and `COMMIT` for the transaction. 

The records you're creating aren't anything problematic, you should be able to insert lots of them within a 20s time limit, which is what leads me to suspect that the transaction is started and then held open waiting from some helper function to run.

>Through the prisma logs I was able to verify that the queries take quite a long time, even if the grouped records are only 3 or 4. Some results were:
>
>Query TokenUsage.createMany took 63444ms (1 record)
Query TokenUsage.createMany took 181053ms (2 record)
Query Record.createMany took 180955ms (3 record)

Just a random guess but those timings are suspiciously close to 1 and 3 minutes. Maybe they match up with timeouts in something else?

**Reply by Abel1011** [reply]
Ok, thanks for the suggestion. I will try the other type of logging and see the results. I will be bringing you the details.

**Reply by Abel1011** [reply]
Hello, thank you very much for your support, after some verifications and code refactoring, the problem has been eliminated and everything is logged normally.

Regards
